nx/lib/nx/defn/expr.ex

```diff
@@ -1182,32 +1191,17 @@ defmodule Nx.Defn.Expr do
     context || acc
   end
 
-  defp maybe_type_shape_string(%{type: type, shape: shape, names: names}) do
-    Nx.Type.to_string(type) <> Nx.Shape.to_string(shape, names)
-  end
-
-  defp maybe_type_shape_string(tuple) when is_tuple(tuple) do
-    list = Tuple.to_list(tuple)
-    IO.iodata_to_binary(["{", Enum.map_intersperse(list, ", ", &maybe_type_shape_string/1), "}"])
-  end
+  require Nx
 
-  defp maybe_type_shape_string(map) when is_map(map) do
-    pairs =
-      Enum.map_intersperse(map, ", ", fn {k, v} ->
-        [inspect(k), " => ", maybe_type_shape_string(v)]
-      end)
-
-    IO.iodata_to_binary(["%{", pairs, "}"])
-  end
-
-  defp maybe_type_shape_string(number) when is_number(number) do
-    shape = {}
-    names = []
-    type = Nx.Type.infer(number)
-    Nx.Type.to_string(type) <> Nx.Shape.to_string(shape, names)
+  defp inspect_as_template(tensor_or_container)
+       when Nx.is_tensor(tensor_or_container) or is_map(tensor_or_container) or
+              is_tuple(tensor_or_container) do
+    Kernel.inspect(Nx.to_template(tensor_or_container),
+      custom_options: [skip_template_backend_header: true]
+    )
   end
```

Comment:
I think this would work too:

```suggestion
  defp inspect_as_template(tensor_or_container) do
    Kernel.inspect(Nx.to_template(tensor_or_container),
      custom_options: [skip_template_backend_header: true]
    )
  end
```

Reply:
It raised for atoms

Reply:
What about:

```
defp inspect_as_template(data) do
  if Nx.Container.impl_for(data) == Nx.Container.Any do
    inspect(data)
  else
    Kernel.inspect(Nx.to_template(tensor_or_container),
      custom_options: [skip_template_backend_header: true]
    )
  end
end
```

?

Reply:
TIL about impl_for

This is much better